src/dune_rules/gen_rules.ml

```diff
@@ -373,38 +373,45 @@ let has_rules subdirs f =
        ; rules
        })
 
-let under_melange_emit_target ~dir =
-  let parent = Path.Build.parent_exn dir in
-  let+ stanzas = Only_packages.stanzas_in_dir parent in
-  match stanzas with
-  | None -> None
-  | Some stanzas ->
-    List.find_map stanzas.stanzas ~f:(function
-      | Melange_stanzas.Emit.T mel ->
-        let target_dir = Path.Build.relative parent mel.target in
-        Option.some_if (Path.Build.equal target_dir dir) mel
-      | _ -> None)
+let rec under_melange_emit_target ~dir =
+  match Path.Build.parent dir with
+  | None -> Memo.return None
+  | Some parent -> (
+    let* stanzas = Only_packages.stanzas_in_dir parent in
+    match stanzas with
+    | None -> under_melange_emit_target ~dir:parent
+    | Some stanzas -> (
+      match
+        List.find_map stanzas.stanzas ~f:(function
+          | Melange_stanzas.Emit.T mel ->
+            let target_dir = Path.Build.relative parent mel.target in
+            Option.some_if (Path.Build.equal target_dir dir) mel
+          | _ -> None)
+      with
+      | Some stanza -> Memo.return @@ Some (parent, stanza)
+      | None -> under_melange_emit_target ~dir:parent))
 
-let melange_emit_rules sctx ~dir stanza =
-  match stanza with
-  | None -> None
-  | Some stanza ->
-    let rules =
-      let parent = Path.Build.parent_exn dir in
-      Rules.collect_unit (fun () ->
-          let* dir_contents = Dir_contents.get sctx ~dir:parent in
-          let* scope = Scope.DB.find_by_dir dir in
-          Melange_rules.setup_emit_js_rules ~dir_contents ~dir ~scope ~sctx
-            stanza)
-    in
-    Some
-      { Build_config.Rules.build_dir_only_sub_dirs = Subdir_set.empty
-      ; directory_targets = Path.Build.Map.empty
-      ; rules
-      }
+let melange_emit_rules sctx stanza ~dir =
+  let rules =
+    Rules.collect_unit (fun () ->
+        let* dir_contents = Dir_contents.get sctx ~dir in
+        let* scope = Scope.DB.find_by_dir dir in
+        Melange_rules.setup_emit_js_rules ~dir_contents ~dir ~scope ~sctx stanza)
+  in
+  { Build_config.Rules.build_dir_only_sub_dirs = Subdir_set.empty
+  ; directory_targets = Path.Build.Map.empty
+  ; rules
+  }
+
+let gen_melange_emit_rules sctx ~dir dir_and_stanza =
+  let open Option.O in
+  let* stanza_dir, (stanza : Melange_stanzas.Emit.t) = dir_and_stanza in
+  if Path.Build.equal (Path.Build.relative stanza_dir stanza.target) dir then
+    Some (melange_emit_rules sctx stanza ~dir:stanza_dir)
+  else None
 
-let gen_melange_emit_rules sctx ~dir stanza =
-  match melange_emit_rules sctx ~dir stanza with
+let gen_melange_emit_rules_or_empty_redirect sctx ~dir dir_and_stanza =
```

Comment:
It is a bit confusing for a reader to understand what's the difference between the dir in `~dir` and the one in `dir_and_stanza`.

Reply:
I added some comments. There's two directories:

1. The one we're loading the rules for
2. The output directory of the emit stanza

Hopefully it should be clear now

Reply:
Thanks, I understand better now. 👌